Live Migration: attach SRIOV devices after migration is finished

ormergi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Following https://github.com/kubevirt/kubevirt/pull/4741
This change handles the SRIOV migration path at the target.

Currently Libvirt does not support migration when host-devices are attached.

On the source side we detach SRIOV devices just before the migration starts.
In order to complete the migration with best effort, we will attach correspond 
SRIOV devices on the target once the migration is finished.

This operation, attaching the SRIOV device on the target, 
does best effort it wont degrade to the VM basic functionality 
in case of a failure. 

This PR also includes:
- Adding new command to virt-launcher cmd-server to handle 
  operations that should be done after a migration is finished.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Rebased on https://github.com/kubevirt/kubevirt/pull/4741

Tested locally (on a SRIOV setup) using #4723.
**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Hot-plug SR-IOV VF interfaces to VM's post a successful migration.
```

ormergi commented:
/cc @EdDev

ormergi commented:
/retest pull-kubevirt-build

ormergi commented:
/test pull-kubevirt-build

EdDev reviewed (changes requested):
Thank you for this PR!

Here some general points that needs a bit of work:
In some cases, the granularity of the commits have been too small to be able and properly review the changes. There have been some cases in this PR that individual functions have been introduced but have not been called yet, which makes the review hard and the story a bit broken.

I would suggest to revisit the flow of the commits and squash the ones that have no existence without the others.
Here is a rough separation (I may have missed some points) option:
- Create the new gRPC command without implementation.
- Populate the implementation (on both client and server) with no-op operations, just to be able and test it and place the infra for the functionality additions.
- Move the functionality of Set*Time to the new command.
- Remove the existing command (Set*Time) from gRPC (anything else should have been removed in the prev step).
- Introduce the "AttachDevice" functionality, in its simple form (i.e. without carrying about not attaching the already attached devices). This should include everything needed for this to work (in sriov, manager and any other place).
- Adjust to only attach the devices that are not yet attached and filter-out the ones that are.
- Adjust the e2e test to remove the XFail.

At each step, some basic reasoning is required. Why a new gRPC command is added, why we move the functionality to it from Set*Time one, why we do that Attach part, why we adjust it. A lot of why :)

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -605,6 +605,39 @@ func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMa
+	currentAttachedHostDevices := domainSpec.Devices.HostDevices
+	sriovDevices, err := sriov.CreateHostDevices(vmi)
+	if err != nil {
+		return []api.HostDevice{}, err
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -605,6 +605,39 @@ func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMa
+	return nil
+}
+
+func getSRIOVHostDevicesToAttach(vmi *v1.VirtualMachineInstance, domainSpec api.DomainSpec) ([]api.HostDevice, error) {
```
Done

ormergi commented:
changes: restructured PR commits according to code review suggestions

ormergi commented:
Changes : rebase to latest of #4741

dhiller commented on tests/vmi_multus_test.go:
```diff
@@ -606,8 +606,7 @@ var _ = Describe("[Serial]SRIOV", func() {
-		sriovcheck := checkSriovEnabled(virtClient, sriovResourceName)
-		if !sriovcheck {
+		if err := validateSRIOVSetup(virtClient, sriovResourceName, 1); err != nil {
 			Skip("Sriov is not enabled in this environment. Skip these tests using - export FUNC_TEST_ARGS='--ginkgo.skip=SRIOV'")
```
Maybe make the `err` part of the Skip message?

ormergi replied:
> Looks good to me in general, two nits inline.

Hi thanks for reviewing this,
the changes suggest mentions are actually from #4741 this PR rebased on.
I prefer not to change these until they merged, we could also those on #4741

dhiller commented on tests/vmi_multus_test.go:
```diff
@@ -1371,21 +1424,26 @@ func configureNodeNetwork(virtClient kubecli.KubevirtClient) {
 	}
-	return false
+	if sriovEnabledNode < minRequiredNodes {
+		return fmt.Errorf("not enough compute nodes with SR-IOV support detected")
```
Maybe make the expected and actual values part of the error message?

dhiller reviewed (commented):
Looks good to me in general, two nits inline.

EdDev reviewed (commented):
> Commit: handler-launcher-com: Add command to handle post-migration

How about adding the server side with an empty function (or some basic thing that can be used later) together with the tests that checks from client to server this command.
I think it will be more complete commit, and the next one will just populate it with content.

Others can even follow this to add other commands.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -1195,7 +1195,8 @@ func (d *VirtualMachineController) migrationTargetExecute(key string,
 			vmiCopy.Status.MigrationState.TargetNodeDomainDetected = true
-			d.setVMIGuestTime(vmi)
+
+			d.finalizeMigration(vmi)
```
I see you updated it, thanks. This can be resolved.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -2424,17 +2425,17 @@ func (d *VirtualMachineController) updateNodeCpuManagerLabel(cpuManagerPath stri
 		return err
 	}
-	err = client.SetVirtualMachineGuestTime(vmi)
+	err = client.FinalizeVirtualMachineMigration(vmi)
```
Thanks, this can be resolved.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -2424,17 +2425,17 @@ func (d *VirtualMachineController) updateNodeCpuManagerLabel(cpuManagerPath stri
+	err = client.FinalizeVirtualMachineMigration(vmi)
 	if err != nil {
-		log.Log.Reason(err).Error("failed to set vmi guest time to the current")
+		log.Log.Reason(err).Error("failed to finalize migration")
```
Can you also add the `vmi` object so the log will include the reference to it.
(I mean `log.Log.Object(vmi).Reason(err)`)

You can do this in a separate commit, up to you.

ormergi replied:
Done

EdDev commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1523,7 +1523,8 @@ var _ = Describe("VirtualMachineInstance", func() {
 		}, 3)
-		It("update guest time after completed migration", func() {
+
+		It("should apply post-migration configurations on guest VM after migration completed", func() {
```
I would replace `configurations` to `operations`, as we do not limit that things we can do there to config stuff.

ormergi replied:
Done

EdDev commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -591,6 +592,19 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
 }
 
+// FinalizeVirtualMachineMigration apply all post-migration changes on the domain.
+func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMachineInstance) error {
```
I think you can resolve this now, it seems to be used now (called).

EdDev reviewed (commented):
> commit: handler-launcher-com: Drop SetVirtualMachineGuestTime command 

A small nit on the commit message, can you just mention the command that replaced it?

EdDev commented on pkg/virt-launcher/virtwrap/cmd-server/server_test.go:
```diff
@@ -211,6 +211,13 @@ var _ = Describe("Virt remote commands", func() {
+			vmi := v1.NewVMIReferenceFromName("testvmi")
+			domainManager.EXPECT().FinalizeVirtualMachineMigration(vmi).Return(nil)
+
+			Expect(client.FinalizeVirtualMachineMigration(vmi)).Should(Succeed())
```
I think you can resolve this.

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -210,3 +223,36 @@ func AttachHostDevices(dom deviceAttacher, hostDevices []api.HostDevice) error {
+	return filteredSlice
+}
+
+func GetSRIOVHostDevicesToAttach(vmi *v1.VirtualMachineInstance, domainSpec api.DomainSpec) ([]api.HostDevice, error) {
```
No need to re-specify "SRIOV", we are under the `sriov` package.

ormergi replied:
Done

ormergi commented:
> > Commit: handler-launcher-com: Add command to handle post-migration
> 
> How about adding the server side with an empty function (or some basic thing that can be used later) together with the tests that checks from client to server this command.
> I think it will be more complete commit, and the next one will just populate it with content.
> 
> Others can even follow this to add other commands.

Done

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -661,10 +661,41 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
 
 // FinalizeVirtualMachineMigration apply all post-migration changes on the domain.
 func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMachineInstance) error {
+	var err error
```
I used it to initilize `virDomain`, well I gurss I need to declare `virDomain` as well so this change will make sense

ormergi replied:
I will remove this along with in-lining the variables you suggests

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -127,14 +127,27 @@ func SafelyDetachHostDevices(domainSpec *api.DomainSpec, eventDetach eventRegist
-	for _, hostDevice := range domainSpec.Devices.HostDevices {
-		if hostDevice.Alias != nil && strings.HasPrefix(hostDevice.Alias.GetName(), AliasPrefix) {
-			hostDevices = append(hostDevices, hostDevice)
+func filterHostDevice(hostDevices []api.HostDevice, filter func(api.HostDevice) bool) []api.HostDevice {
```
I think that filter functions should get a predicate default, 
specifying predicate explicitly might make one think there is are options
What you think?

ormergi replied:
I am removing this decoupling, until we would accurately need it

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -191,3 +191,22 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+			return fmt.Errorf("failed to encode (xml) host-device %v, err: %v", hostDev, err)
+		}
+		err = dom.AttachDevice(string(devXML))
+		if err != nil {
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -191,3 +191,22 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+		if err != nil {
+			return fmt.Errorf("failed to attach host-device %s, err: %v", devXML, err)
+		}
+		log.Log.Infof("Successfully hot-plug host-device: %s (%+v)", hostDev.Alias.GetName(), hostDev.Source.Address)
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -191,3 +191,22 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+		if err != nil {
+			return fmt.Errorf("failed to encode (xml) host-device %v, err: %v", hostDev, err)
+		}
+		err = dom.AttachDevice(string(devXML))
```
I changed it to use the flags version

ormergi replied:
As we know *CURRENT flag will be converted to *LIVE if the domain is running or to *CONFIG if its not.
When  both `*CONFIG` and `*LIVE` flags are used  , libvirt raise the following error:
{"component":"virt-launcher","kind":"","level":"error","msg":"failed to finalize migration","name":"testvmi-zj77c","namespace":"kubevirt-test-default1","pos":"server.go:266","reason":
"failed to attach host-device 
<hostdev type="pci" managed="no">
  <source>
    <address type="pci" domain="0x0000" bus="0x04" slot="0x11" function="0x2"></address>
  </source>
  <alias name="ua-sriov-sriov"></alias>
</hostdev>, 
err: virError(Code=27, Domain=20, Message='XML error: 
non unique alias detected: ua-sriov-sriov')
```

It works when using only `*LIVE` flag (or using `AttachDevice` as before)

ormergi replied:
@EdDev As we know *CURRENT flag will be converted to *LIVE if the domain is running or to *CONFIG if its not.
When  both `*CONFIG` and `*LIVE` flags are used  , libvirt raise the following error:
```
{"component":"virt-launcher","kind":"","level":"error","msg":"failed to finalize migration","name":"testvmi-zj77c","namespace":"kubevirt-test-default1","pos":"server.go:266","reason":
"failed to attach host-device 
<hostdev type="pci" managed="no">
  <source>
    <address type="pci" domain="0x0000" bus="0x04" slot="0x11" function="0x2"></address>
  </source>
  <alias name="ua-sriov-sriov"></alias>
</hostdev>, 
err: virError(Code=27, Domain=20, Message='XML error: 
non unique alias detected: ua-sriov-sriov')
```

It works when using only `*LIVE` flag (or using `AttachDevice` as before)

ormergi replied:
> Is this equivalent to using AttachDeviceFlags with all flags "on"?

I verified and `AttachDevice` is equivalent to `AttachDeviceFlags` with `*LIVE` flag

EdDev replied:
You must have both set, otherwise the persisted config is not updated.

EdDev replied:
Yes, looks like we have a bug in the detach part. We do not persist the operation, I need to fix that.

EdDev replied:
@ormergi , with #5152 it should work for you. Can you please check?

ormergi replied:
I check it and it solve this problem :)

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -661,10 +661,41 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
+	defer virDomain.Free()
+
+	err = hotPlugHostDevices(vmi, virDomain)
+	if err != nil {
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -661,10 +661,41 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
+	}
+
+	err = l.SetGuestTime(vmi)
+	if err != nil {
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -661,10 +661,41 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
+	}
+
+	err = sriov.AttachHostDevices(domain, sriovHostDevices)
+	if err != nil {
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -278,6 +278,37 @@ var _ = Describe("SRIOV HostDevice", func() {
+		})
+
+		It("succeeds to attach more the one device", func() {
+			hostDevice2 := api.HostDevice{Alias: api.NewUserDefinedAlias(sriov.AliasPrefix + "net2")}
```
Done, I added a valid alias

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -278,6 +278,37 @@ var _ = Describe("SRIOV HostDevice", func() {
+			Expect(sriov.AttachHostDevices(obj, []api.HostDevice{hostDevice})).Should(Succeed())
+		})
+
+		It("succeeds to attach more the one device", func() {
```
Fixed

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -210,3 +223,36 @@ func AttachHostDevices(dom deviceAttacher, hostDevices []api.HostDevice) error {
+// it returns a slice with host-devices that exists on 'hostDevices' and not exists on 'removeHostDevices'.
+func FilterOutHostDevicesByAlias(hostDevices, removeHostDevices []api.HostDevice) []api.HostDevice {
+	hostDevices = filterHostDevice(hostDevices, hostDeviceHasAlias)
+	removeHostDevices = filterHostDevice(removeHostDevices, hostDeviceHasAlias)
```
Done

ormergi commented:
Changes: Addressed code review comments
Removed the filtering host-devices decoupling for now

ormergi commented:
We got the impression that libvirt `AttachDevice` is an old API and we should use `AttachDeviceFlags` instead.
Following #5010 domains has persistent config so we need to use `VIR_DOMAIN_AFFECT_CONFIG` flag
and `VIR_DOMAIN_AFFECT_Live` flag to affect the running domain.
When using **both** flags or the `*CONFIG` flag only, libvirt raise the following error
```
{"component":"virt-launcher","kind":"","level":"error","msg":"failed to finalize migration","name":"testvmi-zj77c","namespace":"kubevirt-test-default1","pos":"server.go:266","reason":
"failed to attach host-device 
<hostdev type="pci" managed="no">
  <source>
    <address type="pci" domain="0x0000" bus="0x04" slot="0x11" function="0x2"></address>
  </source>
  <alias name="ua-sriov-sriov"></alias>
</hostdev>, 
err: virError(Code=27, Domain=20, Message='XML error: 
non unique alias detected: ua-sriov-sriov')
```

It works when using only `*LIVE` flag (or using `AttachDevice` as before)
@EdDev

EdDev commented:
/sig network

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -223,3 +223,24 @@ func AttachHostDevices(dom deviceAttacher, hostDevices []api.HostDevice) error {
+
+// FilterOutHostDevicesByAlias given two slices of host-devices, according to Alias.Name,
+// it returns a slice with host-devices that exists on 'hostDevices' and not exists on 'removeHostDevices'.
+func FilterOutHostDevicesByAlias(hostDevices, removeHostDevices []api.HostDevice) []api.HostDevice {
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -191,3 +193,28 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+}
+
+func AttachHostDevices(dom deviceAttacherFlags, hostDevices []api.HostDevice) error {
+	attachDeviceFlags := prepareAttachDeviceFlags()
```
Following https://github.com/kubevirt/kubevirt/pull/5152 I will export it to a constant we could reuse in this package

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -191,3 +193,28 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
 	return nil
 }
+
+type deviceAttacherFlags interface {
```
Got you I will change it and align with deviceDetach stub

ormergi commented on pkg/virt-handler/cmd-client/client.go:
```diff
@@ -464,10 +463,6 @@ func (c *VirtLauncherClient) SyncMigrationTarget(vmi *v1.VirtualMachineInstance)
 
 }
 
-func (c *VirtLauncherClient) SetVirtualMachineGuestTime(vmi *v1.VirtualMachineInstance) error {
```
I removed it one this [commit](https://github.com/kubevirt/kubevirt/pull/5037/commits/effc90f0733776fdbb04983c6d59d7810b4d2d90#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aL2433)

ormergi commented:
Rebase on top of #5152

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -191,3 +193,28 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+		}
+		err = dom.AttachDeviceFlags(string(devXML), attachDeviceFlags)
+		if err != nil {
+			return fmt.Errorf("failed to attach host-device %s, err: %v", devXML, err)
```
I like this idea

ormergi commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -685,7 +685,12 @@ func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMa
+		return err
+	}
+
+	sriovHostDevices, err := sriov.GetHostDevicesToAttach(vmi, *domainSpec)
```
I tried to minimize the side affect of moving a pointer instead of a struct, but I guess that since I pass the VMI pointer its kinda misses the point..

ormergi replied:
I change it to get pointer now

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -24,6 +24,7 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	"libvirt.org/libvirt-go"
```
Done

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -218,3 +218,33 @@ func AttachHostDevices(dom deviceAttacherFlags, hostDevices []api.HostDevice) er
+// DifferenceHostDevicesByAlias given two slices of host-devices, according to Alias.Name,
+// it returns a slice with host-devices that exists on the first slice and not exists on the second.
+func DifferenceHostDevicesByAlias(hostDevices, removeHostDevices []api.HostDevice) []api.HostDevice {
+	lookupHostDeviceToRemove := make(map[string]struct{}, len(removeHostDevices))
```
I changed it to `actualHostDevicesByAlias`

ormergi commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -218,3 +218,33 @@ func AttachHostDevices(dom deviceAttacherFlags, hostDevices []api.HostDevice) er
+
+// DifferenceHostDevicesByAlias given two slices of host-devices, according to Alias.Name,
+// it returns a slice with host-devices that exists on the first slice and not exists on the second.
+func DifferenceHostDevicesByAlias(hostDevices, removeHostDevices []api.HostDevice) []api.HostDevice {
```
Done

ormergi commented:
Changes: addressed code review comments
`AttachDevices` now do best effort when devices fails to attach,
it will attach the rest of the devices and return as error with the devices aliases that failed to attach.

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -193,3 +193,45 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+}
+
+func AttachHostDevices(dom deviceAttacher, hostDevices []api.HostDevice) error {
+	var errors []error
```
nit: I would use `errs` as `errors` is a known std package.

ormergi replied:
That make sense, I changed it

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -193,3 +193,45 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+
+func buildHostPlugHostDevicesErrorMessage(errors []error) error {
+	errorMessageBuilder := strings.Builder{}
+	errorMessageBuilder.WriteString("failed to hot-plug all devices:\n")
```
please drop the `all`, it is confusing.

ormergi replied:
Done

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -280,6 +280,43 @@ var _ = Describe("SRIOV HostDevice", func() {
+			Expect(sriov.AttachHostDevices(obj, []api.HostDevice{hostDevice})).ShouldNot(Succeed())
+		})
+
+		It("error should contain at least that Alias of each device that failed to attach", func() {
```
Why `at least`? and I think you meant `the Alias` (and not `that Alias`)

ormergi replied:
I wrote at least because the device alias is the minimal detail that we want to have in the logs
currently we dump the host-device XML, in case it will change I want this test to fail if a device fail to attach and the log does not contain the device alias.

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -280,6 +280,43 @@ var _ = Describe("SRIOV HostDevice", func() {
+			obj := deviceAttacherFlagsStub{fail: true}
+			hostDevice2 := api.HostDevice{Alias: api.NewUserDefinedAlias("net2")}
+			err := sriov.AttachHostDevices(obj, []api.HostDevice{hostDevice,hostDevice2})
+			Expect(err).ToNot(BeNil())
```
You can wrap `sriov.AttachHostDevices` and use `Succeed`.

ormergi replied:
I reuse the `err` for the assertion on line 315

EdDev replied:
The convention is then to use `HaveOccurred()`.

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -280,6 +280,43 @@ var _ = Describe("SRIOV HostDevice", func() {
+			Expect(err).ToNot(BeNil())
+			Expect(err.Error()).To(And(
+				ContainSubstring(hostDevice.Alias.GetName()),
+				ContainSubstring(hostDevice.Alias.GetName())))
```
I think you meant `hostDevice2`.

ormergi replied:
Thanks I will fix it

EdDev commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go:
```diff
@@ -322,6 +359,17 @@ func (d deviceDetacherStub) DetachDeviceFlags(data string, flags libvirt.DomainD
 	return nil
 }
 
+type deviceAttacherFlagsStub struct {
```
Please remove the `Flags` part as you did in the production code.

ormergi replied:
I forgot to rename it back, I will change it

ormergi commented:
/cc @AlonaKaplan

ormergi commented:
It passed on a local setup, trying again

/test pull-kubevirt-unit-test

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -658,6 +659,18 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
 
+// FinalizeVirtualMachineMigration apply all post-migration changes on the domain.
+func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMachineInstance) error {
+	l.domainModifyLock.Lock()
```
This PR is just moving `SetGuestTime` to be invoked from `FinalizeVirtualMachineMigration` why the extra lock is needed?

ormergi replied:
On following commit this function also handles hostDevices attachment which change the domain so I though it will be a good idea to put a lock.
But now I realize that there is already a lock on setGuestTimeContext.
Should I remove the locks from all the function  `FinalizeVirtualMachineMigration` calls, so it will be handled from one place?
Or each function should be self contained and each function that changes the VM will have a lock?

EdDev replied:
Ohh, I totally missed this one. From some unexplained reason I thought it was something that existed before, but it is added as a new.

A lock here is wrong, it should be placed close to where the domain if fetched and acted on.
`SetGuestTime` had no such lock before, it used a more grained lock for the timer, which looks odd and scary.

The `domainModifyLock` is needed on the "attach" operation (including the reading of the current state), so please add it there.

ormergi replied:
Done, move the lock to `hotPlugHostDevices`

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -106,6 +106,7 @@ type DomainManager interface {
 	GetUsers() ([]v1.VirtualMachineInstanceGuestOSUser, error)
 	GetFilesystems() ([]v1.VirtualMachineInstanceFileSystem, error)
 	SetGuestTime(*v1.VirtualMachineInstance) error
```
Can `SetGuestTime` be moved out of the `DomainManager` and changed to be private?

ormergi replied:
Yes I think we can do that

ormergi replied:
Done

AlonaKaplan commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -193,3 +193,45 @@ func deviceLookup(hostDevices []api.HostDevice, deviceAlias string) *api.HostDev
+	}
+	err = dom.AttachDeviceFlags(string(devXML), libvirt.DOMAIN_DEVICE_MODIFY_LIVE | libvirt.DOMAIN_DEVICE_MODIFY_CONFIG)
+	if err != nil {
+		return fmt.Errorf("failed to attach host-device %s, err: %v", devXML, err)
```
Printing the whole `devXML`  and not just the alias of the device is "by design" or a typo?

>> The error will include the Alias of each device that
>> failed to attach.
I'm asking since that's what was written on the commit message.

ormergi replied:
It should print the device XML, its  small few lines.
I will change the commit message sorry

EdDev reviewed (changes requested):
Please fix the lock issue found by @AlonaKaplan .

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -664,13 +664,39 @@ func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMa
 
+// hotPlugHostDevices attach host-devices to running domain
+// Currently only SRIOV host-devices are supported
+func hotPlugHostDevices(vmi *v1.VirtualMachineInstance, domain cli.VirDomain) error {
```
Since the parallel detach method is called `detachHostDevices` I would call this method `attachHostDevices`.

ormergi replied:
And on other components we call it "hot-plug/unplug" (instead of attach/detach) as this is the general name of connecting/disconnecting a device from a live VM.
At virt-launcher/virtwrap scope I kept the terminology to be "attach/detach" as this is how the operation called on Libvirt

EdDev replied:
I think @AlonaKaplan  has a point here.

> At virt-launcher/virtwrap scope I kept the terminology to be "attach/detach" as this is how the operation called on Libvirt

I guess it is a typo, as at this abstraction level you used "hotPlug".

I used "detach" at the time as this is the terms libvirt used in their API, but we can change it to hotPlug and HotUnplug.
If you think it should change, it makes sense to change it for both cases for consistency.
Consider changing the to hot* in a separate PR and keeping here that same naming.

ormergi replied:
Sorry I meant `virt-launcher/virtwarp/device/sriov package`
We already have functions that called `attachHostDevices` on `virt-launcher/virtwarp/device/sriov` package
so we will end up with `attachHostDevices` calling `sriov.AttachHostDevices`.
I think its make more sense to keep this name, use "attach/detach" naming inside `virt-launcher/virtwarp/device/sriov `package and hotPlug/unplug outside of it.
I prefer to rename `detachHostDevices` on virt-launcher/virtwarp/manager to `hotUnplugHostDevices`
on this PR, there is no need to run all CI just for renaming one function and its still make sense to do it here
What you think?

EdDev replied:
Fine by me.

AlonaKaplan commented on pkg/virt-launcher/virtwrap/device/sriov/hostdev.go:
```diff
@@ -237,3 +237,33 @@ func buildHostPlugHostDevicesErrorMessage(errors []error) error {
 	return fmt.Errorf(errorMessageBuilder.String())
 }
+
+func GetHostDevicesToAttach(vmi *v1.VirtualMachineInstance, domainSpec *api.DomainSpec) ([]api.HostDevice, error) {
```
Is the dom.attach command is blocking?
If not, what if there are some sriov hostdevices that are in the process of being attached?

EdDev replied:
I just confirmed it with @berrange , https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainAttachDeviceFlags is confirmed to be "blocking".

This whole attachment handling should be protected by the `domainModifyLock` lock, something that was already included (but needs to move closer to here).

ormergi commented:
Addressed code review comments

EdDev reviewed (approved):
Looks great, please rebase.

ormergi commented:
rebased

ormergi commented:
/cc @vladikr

EdDev reviewed (commented):
You have a type in the `virt-launcher manager: Rename detachHostDevices` commit.

`naeming`

BTW, the subject in the commit is just `virt-launcher manager: Rename detachHostDevices` ?

ormergi commented:
> You have a typo in the `virt-launcher manager: Rename detachHostDevices` commit.
> 
> `naeming`
>
Done

> BTW, the subject in the commit is just `virt-launcher manager: Rename detachHostDevices` ?

Yes, do you think there should be more to it?

EdDev commented:
> > BTW, the subject in the commit is just `virt-launcher manager: Rename detachHostDevices` ?
> 
> Yes, do you think there should be more to it?

No, it is fine. I just was not sure if what GitHub shows me is indeed the case.

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -663,13 +663,42 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
+	defer virDomain.Free()
+
+	if err := l.hotPlugHostDevices(vmi, virDomain); err != nil {
+		return err
```
Since the hotplug is "nice to have" (the migration doesn't fail in case the some of the attachements failed), I believe we shouldn't return an error here and setting the `GuestTime` even in case the `hotPlugHostDevices` had an error.

ormergi replied:
I changed it so it will log the error instead

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -663,13 +663,42 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
 // FinalizeVirtualMachineMigration apply all post-migration changes on the domain.
 func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMachineInstance) error {
+	domainName := api.VMINamespaceKeyFunc(vmi)
+	virDomain, err := l.virConn.LookupDomainByName(domainName)
```
Reading the domain is for the hotplug only, so I believe it should be part of `hotPlugHostDevices` and after locking `domainModifyLock` (isn't that the whole purpose on this lock?).

ormergi replied:
I moved "reading domain" part to `hotPlugHostDevices`

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -663,13 +663,42 @@ func (l *LibvirtDomainManager) getGuestTimeContext() context.Context {
+	}
+	defer domain.Free()
+
+	l.domainModifyLock.Lock()
```
Can you please explain why `l.virConn.LookupDomainByName` doesn't need to be locked as well?

ormergi replied:
I though that `l.domainModifyLock` should be locked when there is an operation that changes the domain
but according to the [comment at its definition](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/manager.go#L115-L116) it should be locked also from "get" operations.

EdDev replied:
It actually does not need to be locked, as it does not change anything and the information it fetches is just a descriptor to the domain. We only have one, and if something is going terribly wrong, nothing will work anyway.

But for simplicity here, I think it is fine to include it in the lock.

ormergi replied:
I think there could be a case where during fetch domain it changes by another consumer
Why its not mandatory to lock on fetch domain operation?

EdDev replied:
What will change? Give me an example.

ormergi replied:
A user desided to rename the VM right after migration is finished and its time to attach host-devices

> /approve

Thank you :slightly_smiling_face:

ormergi commented:
Rebased

ormergi commented:
Rebased

AlonaKaplan commented:
/approve

ormergi commented:
> /approve

Thanks you :slightly_smiling_face:

EdDev reviewed (approved):
Great work, thank you!

ormergi commented:
> Great work, thank you!

Thanks from all the reviews :slightly_smiling_face:
